### PR TITLE
Rework allocator sharing to work for multiple devices.

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Rework the allocator sharing infrastructure to work for all device types.
Simplify to make more device centric. 

**Motivation and Context**
Enable sharing CPU allocator between more CPU based EPs

### CODE REVIEW

File: include/onnxruntime/core/framework/execution_provider.h [hunk 1]
```diff
@@ -258,25 +257,24 @@ class IExecutionProvider {
...
 
   /**
-     Register allocators used for EP
-     TODO: Used for CUDA & TRT only for now, will have one more PR to apply this for all EPs.
-     EPs will have a shared pointer to allocator_manager, allocator_managerall will be the only place for allocators
+     Register allocators for EP, potentially re-using existing allocators for a device from allocator_manager.
+     If the EP implements this it should delay creating any allocators until this is called.
   */
-  virtual void RegisterAllocator(std::shared_ptr<AllocatorManager> allocator_manager);
+  virtual void RegisterAllocator(AllocatorManager& /*allocator_manager*/);
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/execution_provider.h:278]: EP doesn't need any ownership of the AllocatorManager.

File: onnxruntime/core/framework/execution_provider.cc [hunk 2]
```diff
@@ -52,103 +52,64 @@
...
-  }
-
-  return false;
-}
-
 // Update allocator in the provider if already present; ignore if not.
+// We match using the device id, OrtMemType and OrtDevice info.
+// We ignore the allocator name, and OrtAllocatorType (whether internally an arena is used or not).
+// TODO: We should remove OrtAllocatorType from OrtMemoryInfo as it's an implmentation detail of the allocator.
 void IExecutionProvider::ReplaceAllocator(AllocatorPtr allocator) {
```
Comment (skottmckay) [inline onnxruntime/core/framework/execution_provider.cc:59]: @pranavsharma let me know if you think there are any issues with this more OrtDevice centric approach.

File: onnxruntime/core/framework/execution_provider.cc [hunk 3]
```diff
@@ -52,103 +52,64 @@
...
-  }
+  auto iter = allocators_.find(MakeKey(info.id, info.mem_type));
+  if (iter != allocators_.end()) {
+    // check device as mem_type is relative to the device
+    // e.g. OrtMemTypeDefault is CPU for a CPU EP and GPU for a CUDA EP. An individual EP will only have one
+    // allocator for an OrtMemType value, so this check is to ensure we don't replace with an incompatible allocator.
+    if (iter->second->Info().device == info.device) {
+      IAllocator* existing_alloc = iter->second.get();
+      for (auto& entry : allocator_list_) {
+        if (entry.get() == existing_alloc) {
```
Comment (skottmckay) [inline onnxruntime/core/framework/execution_provider.cc:70]: allocator_list_ wasn't previously updated but should be in sync with the values in allocators_

File: onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc [hunk 4]
```diff
@@ -161,10 +169,12 @@
...
         },
-        DEFAULT_CPU_ALLOCATOR_DEVICE_ID);
+        cpu_device.Id());
+
     hip_cpu_alloc = CreateAllocator(cpu_memory_info);
-    allocator_manager->InsertAllocator(hip_cpu_alloc);
+    allocator_manager.InsertAllocator(hip_cpu_alloc);
   }
+
   TryInsertAllocator(hip_cpu_alloc);
```
Comment (skottmckay) [inline onnxruntime/core/providers/migraphx/migraphx_execution_provider.cc:178]: All changes in this file below here are from clang-format

File: onnxruntime/test/providers/tensorrt/tensorrt_basic_test.cc [hunk 5]
```diff
@@ -106,7 +106,7 @@ void RunWithOneSessionSingleThreadInference(std::string model_name, std::string
   RunOptions run_options;
   run_options.run_tag = so.session_logid;
   InferenceSession session_object{so, GetEnvironment()};
-  auto allocator_manager = session_object.GetAllocatorManager();
+  onnxruntime::AllocatorManager allocator_manager;
   auto cuda_provider = DefaultCudaExecutionProvider();
```
Comment (skottmckay) [inline onnxruntime/test/providers/tensorrt/tensorrt_basic_test.cc:110]: AllocatorManager is just a container that is temporarily used to share allocators while calling RegisterAllocator. We don't need the AllocatorManager instance from the session, so removed GetAllocatorManager from that API as it was only used by these tests..

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 6]
```diff
@@ -2431,35 +2431,46 @@ CUDAExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
...
-  auto cuda_alloc = allocator_manager->GetAllocator(info_.device_id, OrtMemTypeDefault);
+  auto cuda_alloc = allocator_manager.GetAllocator(OrtMemTypeDefault, cuda_device);
   if (nullptr == cuda_alloc) {
     cuda_alloc = CreateCudaAllocator(info_.device_id, info_.gpu_mem_limit, info_.arena_extend_strategy,
                                      info_.external_allocator_info, info_.default_memory_arena_cfg);
-    allocator_manager->InsertAllocator(cuda_alloc);
+    allocator_manager.InsertAllocator(cuda_alloc);
   }
+
   TryInsertAllocator(std::move(cuda_alloc));
```
Comment (skottmckay) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:2448]: I don't understand why we need TryInsertAllocator in these EPs. This should be the only place we're inserting an allocator into the per-EP list, and RegisterAllocator is only called once per EP in the inference session. Is there a scenario where we need this? 

Only one that comes to mind is if someone registers two instances of the same EP with the same device_id (as the device info is part of the key).

File: onnxruntime/core/session/inference_session.h [hunk 7]
```diff
@@ -780,7 +776,7 @@ class InferenceSession {
   // "session.use_ort_model_bytes_directly" to "1", this will be empty
   std::vector<uint8_t> ort_format_model_bytes_data_holder_;
 
-  std::shared_ptr<onnxruntime::AllocatorManager> allocator_manager_;
+  AllocatorManager allocator_manager_;
```
Comment (snnn) [inline onnxruntime/core/session/inference_session.h:779]: An EP could live longer than a session. So, what if an EP is still alive but the allocator manager is gone?

File: onnxruntime/core/framework/allocatormgr.h [hunk 8]
```diff
@@ -40,24 +36,20 @@ struct AllocatorCreationInfo {
...
   void InsertAllocator(AllocatorPtr allocator);
   void ReplaceAllocator(AllocatorPtr allocator);
-  //Get an allocator with specified device id and MemType. Return nullptr if it doesn't exist
-  AllocatorPtr GetAllocator(int id, OrtMemType mem_type) const;
+  // Get an allocator for the device. Return nullptr if it doesn't exist
+  AllocatorPtr GetAllocator(OrtMemType mem_type, OrtDevice device) const;
 
  private:
+  // key from OrtMemType+OrtDevice mapped to AllocatorPtr
+  using AllocatorMap = std::unordered_map<int32_t, AllocatorPtr>;
```
Comment (snnn) [inline onnxruntime/core/framework/allocatormgr.h:51]: So,  this class, even it is called "manager", doesn't actually own the allocator?

**Comment by snnn** [comment]
LGTM. Please resolve the conflict.

**Comment by pranavsharma** [comment]
I'll take a look shortly.

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 6 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc:2448]: The CIs I tried with InsertAllocator passed, so using that unless there's a need to support the 'try' approach.

**Comment by skottmckay** [comment]
> I'll take a look shortly.

Thanks. 

One note on the EP interface change. It only affects EPs that are attempting allocator sharing. Not sure if that applies to any external EPs that are not in this repo. I assumed given the existing sharing was non-cpu only that this was highly unlikely.

**Review by pranavsharma — APPROVED** [review]

File: onnxruntime/core/framework/allocatormgr.cc [hunk 9]
```diff
@@ -62,30 +68,27 @@
 // Update allocator in the provider if already present; ignore if not.
 void AllocatorManager::ReplaceAllocator(AllocatorPtr allocator) {
   const auto& info = allocator->Info();
-  auto ite = mem_info_set_.find(info);
-  if (ite != mem_info_set_.end()) {
-    const int key = MakeKey(info.id, info.mem_type);
-    allocators_[key] = allocator;
+  auto iter = allocators_.find(MakeKey(info.mem_type, info.device));
+  if (iter != allocators_.end()) {
+    iter->second = allocator;
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/allocatormgr.cc:73]: >allocator [](http://example.com/codeflow?start=19&length=9)

std::move()

File: onnxruntime/core/framework/allocatormgr.cc [hunk 10]
```diff
@@ -62,30 +68,27 @@
...
+    ORT_THROW("Duplicate allocator for OrtMemType:", info.mem_type, " device:", info.device.ToString(),
+              " Existing allocator: ", iter->second->Info().name,
+              " New allocator: ", allocator->Info().name);
   }
-  const int key = MakeKey(info.id, info.mem_type);
-  allocators_.insert({key, allocator});
-  mem_info_set_.insert(ite, info);
-  allocator_list_.push_back(allocator);
+
+  allocators_[key] = allocator;
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/allocatormgr.cc:87]: >allocator [](http://example.com/codeflow?start=21&length=9)

std::move()

File: onnxruntime/core/framework/execution_provider.cc [hunk 11]
```diff
@@ -52,103 +52,52 @@
...
-                                allocators_, allocator)) {
-      // We found an allocator corresponding to the mutated OrtMemoryInfo
-      // and we replaced it with the provided allocator.
-      // Before we return back, we need to do some house-keeping
-      // (i.e.) update the EP's OrtMemoryInfo set
-
-      // Delete the existing OrtMemoryInfo  corresponding to the allocator
-      // that was replaced
-      mem_info_set_.erase(check_info);
+  // TODO: This only works on allocators that are stored in this class. If a derived class overrides GetAllocator
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/execution_provider.cc:62]: >If a derived class overrides GetAllocato [](http://example.com/codeflow?start=72&length=40)

Should we make a comment about it on the class level or even make the class final?

File: onnxruntime/core/framework/execution_provider.cc [hunk 12]
```diff
@@ -52,103 +52,52 @@
...
+  auto iter = allocators_.find(MakeKey(info.id, info.mem_type));
+  if (iter != allocators_.end()) {
+    // check device as mem_type is relative to the device
+    // e.g. OrtMemTypeDefault is CPU for a CPU EP and GPU for a CUDA EP. An individual EP will only have one
+    // allocator for an OrtMemType value, so this check is to ensure we don't replace with an incompatible allocator.
+    if (iter->second->Info().device == info.device) {
+      IAllocator* existing_alloc = iter->second.get();
+      for (auto& entry : allocator_list_) {
+        if (entry.get() == existing_alloc) {
+          entry = allocator;
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/execution_provider.cc:75]: >allocator [](http://example.com/codeflow?start=18&length=9)

move?

File: onnxruntime/core/framework/session_state.h [hunk 13]
```diff
@@ -408,20 +408,29 @@ class SessionState {
...
+  // currently the allocator type is an implementation detail and we don't make any behavioral choices based on it,
   // so exclude it from the key comparison for allocator_idx_map_.
   // we also don't expect to have two allocators with the same name, one using an arena and one not.
-  struct OrtMemoryInfoLessThanIgnoreAllocType {
+  struct OrtMemoryInfoLessThanIgnoreNameAndAllocType {
     bool operator()(const OrtMemoryInfo& lhs, const OrtMemoryInfo& rhs) const {
-      //if (lhs.alloc_type != rhs.alloc_type)
-      //  return lhs.alloc_type < rhs.alloc_type;
+      // if (lhs.alloc_type != rhs.alloc_type)
+      //   return lhs.alloc_type < rhs.alloc_type;
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/session_state.h:417]: Should we remove this?

**Comment by yuslepukhin** [comment]
>   // using std::map as OrtMemoryInfo would need a custom hash function to be used with std::unordered_map,

If this is converted into a hash_map, we would need to ignore allocator type in a hash_function

---
Refers to: onnxruntime/core/framework/session_state.h:437 in 9675c0d. [](commit_id = 9675c0dcbdcd06d385d7afc32d955bde553d50f9, deletion_comment = False)

**Review by yuslepukhin — APPROVED** [review]
:shipit:

**Review by pranavsharma — APPROVED** [review]

**Pull request merged**